vermouth/processors/molecule_grouper.py

```diff
@@ -0,0 +1,295 @@
+    data_nodes = list(range(len(data)))
+    clust_nodes = list(range(len(data), len(data) + num_clusters))
+    artificial_node = len(data) + num_clusters
+    # Demand for artificial node
+    rest_cost = len(data) - sum(clust_sizes)
+    flow_graph.nodes[artificial_node]['demand'] = rest_cost
+    for clust_n_idx in clust_nodes:
+        # Edges between clusters and artificial node
+        # The paper only has an edge from clusters to artificial node with an
+        # infinite capacity and weight 0, since it only deals with a *minimal*
+        # number of items per cluster (i.e. you can add more items to any
+        # cluster to satisfy the cluster's requirements, and shove any
+        # superfluous items on to the artificial node).
+        # The capacity means the cluster can donate (or receive) a limited
+        # number of supply from the artificial node, and this number exactly is
+        # the deviation from the desired number of items per cluster.
+        # The weight affects how often/easy it is to deviate from the required
+        # number of items per cluster, even when a perfect/equal partitioning is
+        # possible.
+        flow_graph.add_edge(clust_n_idx, artificial_node,
+                            weight=round(precision*violation_cost**2),
+                            capacity=tolerances[clust_n_idx-len(data_nodes)])
+        flow_graph.add_edge(artificial_node, clust_n_idx,
+                            weight=round(precision*violation_cost**2),
+                            capacity=tolerances[clust_n_idx-len(data_nodes)])
+        # Demand for cluster nodes
+        flow_graph.nodes[clust_n_idx]['demand'] = clust_sizes[clust_n_idx - len(data)]
+    for data_n_idx in data_nodes:
+        # Data supply
+        flow_graph.nodes[data_n_idx]['demand'] = -1
+    for data_n_idx, clust_n_idx in itertools.product(data_nodes, clust_nodes):
+        # Edges between data and clusters
+        # Capacity is meaningless, since data nodes have only 1 supply to give,
+        # but it seems to help the network simplex a little.
+        flow_graph.add_edge(data_n_idx, clust_n_idx, capacity=1)
+
+    # Start clustering iterations.
+    iter = 0
+    flow_dict = {}
+    while iter < max_iter:
+        # Step 1: set up edge weights based on clustroid positions
+        # TODO: PBC
+        dists = np.sum((data[:, np.newaxis] - clusters[np.newaxis, :])**2, axis=-1)
+        # When clusters are empty their clustroids become nan, which we can't
+        # round to an int.
+        dists[np.isnan(dists)] = 10**2
+        for (data_idx, clust_idx), dist in np.ndenumerate(dists):
+            # Network simplex (and capacity scaling) does not work with float
+            # weights according to nx documentation. So round the distances to
+            # an appropriate int. Magic.
+            flow_graph.edges[data_idx, clust_idx + len(data)]['weight'] = round(precision*dist)
```

Comment:
To what extend to your results relay on selecting a good precision? I'm asking because you only tried charmm tip3 water? If I would use SPC or so that would change anything would it? Density and coordination specifics can be different especially at non-room temperatures.

Reply:
Untested, but I don't expect it to matter too much. The only reason for the precision argument to be there is that the network simplex can only deal with integers, so all floats have to be rounded to an int. The `precision` determines the precision of that rounding (although perhaps it's semantics should be inverted, so that a precision of `1e-8` means up to 8 decimals are taken into account). 
Bear in mind that the input file format also has a limited precision. Going over that will be pointless anyway.